### ISSUE TITLE
Why I would use the houlsby adapter instead of the pfeiffer one?

### ISSUE THREAD

**Issue by piegu** [adapter-hub/adapters#168]
Hi,

I searched a fair comparaison between the `houlsby` ([Houlsby et al. (2019) a](http://proceedings.mlr.press/v97/houlsby19a/houlsby19a.pdf)) and `pfeiffer` ([Pfeiffer et al. (2020a)](https://arxiv.org/pdf/2005.00247.pdf)) adapters. 

In the page 14 of "[AdapterFusion: Non-Destructive Task Composition for Transfer Learning](https://arxiv.org/pdf/2005.00247.pdf)", I found it:

`We find that in contrast to Houlsby et al. (2019), but in line with Bapna and Firat (2019), a single adapter after the feed-forward layer outperforms other settings. While we find that this setting performs on-par with that of Houlsby et al. (2019), it requires only half the number of newly introduced adapters as compared to them, resulting in a more efficient setting in terms of number of operations.`

According to this text, it is useless to use the `houlsby` adapter.

So, why in the [adapter-transformers](https://github.com/Adapter-Hub/adapter-transformers) library and in the [AdapterHub](https://adapterhub.ml/), I can choose `houlsby` adapter?

**Obs**: in order to get a visual comparaison between the 2 adapters, see page 5 of [AdapterHub: A Framework for Adapting Transformers](https://arxiv.org/pdf/2007.07779.pdf)

**Comment by piegu** [comment]
Thanks @JoPfeiff for your answer (and for your great work with Adapter-Transformers!).

You wrote: `not training an adapter after the FF in the last layer results in considerable performance gains`

Do you plan to publish soon the corresponding code? Thanks.

**Comment by piegu** [comment]
> this means that there are 0 adapters in the last transformer layer

Hello @JoPfeiff. You call this **MAD-X 2.0** in your paper [UNKs Everywhere: Adapting Multilingual Language Models to New Scripts](https://arxiv.org/pdf/2012.15562.pdf) but I would like to know if you only talk about the task adapter here (when you say 0 adapters in the last transform layer), that would mean that you keep the Language Adapter in the last transform layer, or if you are talking about both Language and Task Adapters that would not appear in the last transform layer in the MAD-X 2.0 approach?

More, do you plan to update the Adapter-Transformers library to release the MAX-X 2.0 code? Thanks.

**Comment by JoPfeiff** [comment]
both adapters (task *and* language) are dropped from the last layer

**Comment by piegu** [comment]
> Potentially having only an adapter after the MHA but not after the FF (only) in the last layer might thus improve performance even more. I have not experimented with this yet though, it's just a hypothesis...

Hello @JoPfeiff. Did you have the time to test this configuration with `houlsby`? 

(until now, it seems not possible to test it with the library `adapter-transformers` that allows to keep all adapters  or nothing in the last layer, but not only an adapter after the MHA but not after the FF)

**Comment by piegu** [comment]
> The "houlsby" architecture does not have this issue, as the adapter component after the MHA would still be encapsulated between the feed-forward layer.
> Potentially having only an adapter after the MHA but not after the FF (only) in the last layer might thus improve performance even more. I have not experimented with this yet though, it's just a hypothesis...

Hello @JoPfeiff. I published a [notebook](https://github.com/piegu/language-models/blob/master/adapters/language-modeling/language_modeling_adapter.ipynb) and a [post](https://medium.com/@pierre_guillou/nlp-nas-empresas-como-ajustar-um-modelo-de-linguagem-natural-como-bert-a-um-novo-dom%C3%ADnio-23752b73b185) (in Portuguese) where I did implement `pfeiffer+inv MAD-X 2.0` and `houlsby+inv without module after MHA`. 

I tested these configurations on a BERT base in Portuguese (and all others: pfeiffer, pfeiffer+inv, houlsby, houlsby+inv).

The results show that `houlsby+inv without module after MHA` has the best performance ... but the `pfeiffer+inv MAD-X 2.0` configuration is very very close. As `pfeiffer+inv MAD-X 2.0` has half the parameters of the `houlsby+inv without module after MHA`, the choice is easy in terms of using adapters in production :-) 

Thank you very much for your work and your `adapter-transformers` library!
